_docs/dsl.md

```diff
@@ -0,0 +1,337 @@
+
+Here is a small yet useful, example of ruleguard file:
+
+```go
+package gorules
+
+import "github.com/quasilyte/go-ruleguard/dsl"
+
+func regexpMust(m dsl.Matcher) {                          //        - regexpMust matcher func
+	m.Match(`regexp.Compile($pat)`,                       // - rule  | (or "regexpMust rules group")
+		`regexp.CompilePOSIX($pat)`).                     //  |      |
+		Where(m["pat"].Const).                            //  |      |
+		Report(`can use MustCompile for const patterns`). //  |      |
+		Suggest(`regexp.MustCompile($pat)`)               // -       |
+}                                                         //        -
+```
+
+A `Report()` argument string can use `$<varname>` notation to interpolate the named pattern submatches into the report message.
+
+There is a special variable `$$` which can be used to inject the entire pattern match into the message (like `$0` in regular expressions).
+
+### Filters
+
+The rule is matched if:
+
+1. At least 1 AST pattern from `Match()` is matched
+2. Filters from `Where()` accept the given match
+
+There are 2 types of filters that can be used in [`Where()`](https://godoc.org/github.com/quasilyte/go-ruleguard/dsl#Matcher.Where) call:
+
+1. Submatch (named variable-based) filters
+2. Context filters (current file, etc)
+
+A match variable describes a named submatch of a pattern.
+
+Here are some examples of supported filters:
+
+* Submatch expression type is identical to `T`
+* Submatch expression type is assignable to `T`
+* Submatch expression type implements interface `I`
+* Submatch expression is side-effect free
+* Submatch expression is a const expression
+* Submatch expression const value check
+* Submatch text matches provided regexp
+* Current files imports package `P`
+
+A match variable can be accessed with `dsl.Matcher` function argument indexing:
+
+```go
+// m["a"] -- $a
+// m["b"] -- $b
+Where(m["a"].Type.Is(`int`) && !m["b"].Type.AssignableTo(`[]string`))
+```
+
+If we had a pattern with `$a` and `$b` match variables, a filter above would only accept it
+if `$a` expression had a type of `int` while `$b` is anything that is **not** assignable to `[]string`.
+
+Context-related filters can be applied through `m` members:
+```go
+// Using m.File() to apply a file-related filter.
+Where(m.File().Imports("io/ioutil"))
+```
+
+The filter concept is crucial to avoid false-positives in rules.
+
+Please refer to the godoc page of a [`dsl`](https://godoc.org/github.com/quasilyte/go-ruleguard/dsl) package to get an up-to-date list of supported filters.
+
+## Custom filters
+
+When none of the DSL filters seem to do what you want, you can write a custom filter function.
+
+```go
+package gorules
+
+import (
+	"github.com/quasilyte/go-ruleguard/dsl"
+	"github.com/quasilyte/go-ruleguard/dsl/types"
+)
+
+func implementsStringer(ctx *dsl.VarFilterContext) bool {
+	stringer := ctx.GetInterface(`fmt.Stringer`)
```

Comment:
It may be worth having an example where the interface name requires an import. Presumably the import does not (and cannot) be done in this function; it needs to be done in the rule right before `m.Match`, for every rule that uses `implementsStringer`

Reply:
The `GetInterface()` (and `GetType()`) functions accept a `FQN`. So it can be a `github.com/a/b.Type` without `Import`. I also made [`Implements` in DSL accept FQN](https://github.com/quasilyte/go-ruleguard/pull/187); so `Import` is mostly needed for the brevity now.

Reply:
It accepts a `typeName` parameter: https://github.com/quasilyte/go-ruleguard/blob/c73ee1cbff1ff0af49b54bcb3d480aa18e1c41cb/dsl/dsl.go#L166-L169